<a href="https://colab.research.google.com/github/pingao2019/DS-Unit-2-Linear-Models/blob/master/Copy_of_h212Copy_of_LS_DS_212_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lambda School Data Science

*Unit 2, Sprint 1, Module 2*

---

# Regression 2

## Assignment

You'll continue to **predict how much it costs to rent an apartment in NYC,** using the dataset from renthop.com.

- [ ] Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
- [ ] Engineer at least two new features. (See below for explanation & ideas.)
- [ ] Fit a linear regression model with at least two features.
- [ ] Get the model's coefficients and intercept.
- [ ] Get regression metrics RMSE, MAE, and $R^2$, for both the train and test data.
- [ ] What's the best test MAE you can get? Share your score and features used with your cohort on Slack!
- [ ] As always, commit your notebook to your fork of the GitHub repo.


#### [Feature Engineering](https://en.wikipedia.org/wiki/Feature_engineering)

> "Some machine learning projects succeed and some fail. What makes the difference? Easily the most important factor is the features used." — Pedro Domingos, ["A Few Useful Things to Know about Machine Learning"](https://homes.cs.washington.edu/~pedrod/papers/cacm12.pdf)

> "Coming up with features is difficult, time-consuming, requires expert knowledge. 'Applied machine learning' is basically feature engineering." — Andrew Ng, [Machine Learning and AI via Brain simulations](https://forum.stanford.edu/events/2011/2011slides/plenary/2011plenaryNg.pdf) 

> Feature engineering is the process of using domain knowledge of the data to create features that make machine learning algorithms work. 

#### Feature Ideas
- Does the apartment have a description?
- How long is the description?
- How many total perks does each apartment have?
- Are cats _or_ dogs allowed?
- Are cats _and_ dogs allowed?
- Total number of rooms (beds + baths)
- Ratio of beds to baths
- What's the neighborhood, based on address or latitude & longitude?

## Stretch Goals
- [ ] If you want more math, skim [_An Introduction to Statistical Learning_](http://faculty.marshall.usc.edu/gareth-james/ISL/ISLR%20Seventh%20Printing.pdf),  Chapter 3.1, Simple Linear Regression, & Chapter 3.2, Multiple Linear Regression
- [ ] If you want more introduction, watch [Brandon Foltz, Statistics 101: Simple Linear Regression](https://www.youtube.com/watch?v=ZkjP5RJLQF4)
(20 minutes, over 1 million views)
- [ ] Add your own stretch goal(s) !

In [0]:
%%capture
import sys

# If you're on Colab:
if 'google.colab' in sys.modules:
    DATA_PATH = 'https://raw.githubusercontent.com/LambdaSchool/DS-Unit-2-Applied-Modeling/master/data/'
    !pip install category_encoders==2.*

# If you're working locally:
else:
    DATA_PATH = '../data/'
    
# Ignore this Numpy warning when using Plotly Express:
# FutureWarning: Method .ptp is deprecated and will be removed in a future version. Use numpy.ptp instead.
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning, module='numpy')

In [0]:
import numpy as np
import pandas as pd

# Read New York City apartment rental listing data
df = pd.read_csv(DATA_PATH+'apartments/renthop-nyc.csv')
assert df.shape == (49352, 34)

# Remove the most extreme 1% prices,
# the most extreme .1% latitudes, &
# the most extreme .1% longitudes
df = df[(df['price'] >= np.percentile(df['price'], 0.5)) & 
        (df['price'] <= np.percentile(df['price'], 99.5)) & 
        (df['latitude'] >= np.percentile(df['latitude'], 0.05)) & 
        (df['latitude'] < np.percentile(df['latitude'], 99.95)) &
        (df['longitude'] >= np.percentile(df['longitude'], 0.05)) & 
        (df['longitude'] <= np.percentile(df['longitude'], 99.95))]

In [4]:
print(df.shape)
df.head(2)

(48817, 34)


bathrooms  bedrooms  ... wheelchair_access common_outdoor_space
0        1.5         3  ...                 0                    0
1        1.0         2  ...                 0                    0

[2 rows x 34 columns]

In [0]:
df.isnull().sum()

In [0]:
df.columns.to_list()

 Do train/test split. Use data from April & May 2016 to train. Use data from June 2016 to test.
Engineer at least two new features. (See below for explanation & ideas.)
Fit a linear regression model with at least two features.

In [0]:
import numpy as np
import datetime

In [5]:

df['month_year'] = pd.to_datetime(df['created']).dt.to_period('M')
df.head(2)

bathrooms  bedrooms  ... common_outdoor_space month_year
0        1.5         3  ...                    0    2016-06
1        1.0         2  ...                    0    2016-06

[2 rows x 35 columns]

In [0]:
df['year'] = pd.DatetimeIndex(df['created']).year

In [0]:
df['month'] = pd.DatetimeIndex(df['created']).month


In [8]:
df['month'].dtypes

dtype('int64')

In [9]:
condtion = (df['year']==2016)
df=df[condtion]
df.head(2)

bathrooms  bedrooms              created  ... month_year  year  month
0        1.5         3  2016-06-24 07:54:24  ...    2016-06  2016      6
1        1.0         2  2016-06-12 12:19:27  ...    2016-06  2016      6

[2 rows x 37 columns]

In [0]:
condtion= (df['month']==5) 
df1=df[condtion]


In [11]:
condtion= (df['month']==4) 
df2=df[condtion]
train = pd.concat([df1, df2])
print(train.shape)
train.tail()

(31844, 37)


bathrooms  bedrooms              created  ... month_year  year  month
49346        1.0         1  2016-04-22 15:44:11  ...    2016-04  2016      4
49348        1.0         1  2016-04-04 18:22:34  ...    2016-04  2016      4
49349        1.0         1  2016-04-16 02:13:40  ...    2016-04  2016      4
49350        1.0         0  2016-04-08 02:13:33  ...    2016-04  2016      4
49351        1.0         2  2016-04-12 02:48:07  ...    2016-04  2016      4

[5 rows x 37 columns]

In [12]:
condtion= (df['month']==6) 
test=df[condtion]
print(test.shape)
test.head()

(16973, 37)


bathrooms  bedrooms              created  ... month_year  year  month
0         1.5         3  2016-06-24 07:54:24  ...    2016-06  2016      6
1         1.0         2  2016-06-12 12:19:27  ...    2016-06  2016      6
11        1.0         1  2016-06-03 03:21:22  ...    2016-06  2016      6
14        1.0         1  2016-06-01 03:11:01  ...    2016-06  2016      6
24        2.0         4  2016-06-07 04:39:56  ...    2016-06  2016      6

[5 rows x 37 columns]

Engineer  three new features.  Fit a linear regression model with two features.

In [13]:
import plotly.express as px
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
model = LinearRegression()
features = ['bedrooms', 'latitude','longitude']
X_train = train[features]
X_test = test[features]
print(f'Linear Regression, dependent on: {features}')
target= 'price'
y_train= train[target]
y_test = test[target]

Linear Regression, dependent on: ['bedrooms', 'latitude', 'longitude']


In [14]:
model.fit(X_train, y_train)
y_pred = model.predict(X_train)
mae = mean_absolute_error(y_train, y_pred)
print(f'Train Error: {mae:.2f} dollars')




Train Error: 883.71 dollars


In [16]:
# apply the model to test data
y_pred = model.predict(X_test)
mae = mean_absolute_error(y_test, y_pred)
print(f'Test Error: {mae:.2f} dollars')


Test Error: 898.31 dollars


Get the model's coefficients and intercept.
Get regression metrics RMSE, MAE, and R2, for both the train and test data.
What's the best test MAE you can get? Share your score and features used with your cohort on Slack!

In [17]:
model.intercept_, model.coef_




(-1453990.5652711, array([   881.61378123,   2593.5292654 , -18257.10644491]))

In [18]:
print('Intercept', model.intercept_)
coefficients = pd.Series(model.coef_, features)
print(coefficients.to_string())

Intercept -1453990.5652711
bedrooms       881.613781
latitude      2593.529265
longitude   -18257.106445


In [22]:
beta0 = model.intercept_
beta1, beta2, beta3 = model.coef_
print(f'y = {beta0} + {beta1}x1 + {beta2}x2+ {beta3}x3')

y = -1453990.5652711 + 881.6137812307751x1 + 2593.5292654049754x2+ -18257.10644491425x3


We got regression equation (price with 3 features: bedrooms, latitude, longitude) like this :  $y = \beta_0 + \beta_1x_1 + \beta_2x_2+ \beta_3x_3 $

$y = -1453990.5652711 + 881.613781x_1 + 2593.529265x_2 -18257.106445x_3 $


In [19]:
# We use this to do prediction. What does the model predict if bedroom= 1, latitude=40.7296, longitude	-73.9869
model.predict([[1, 40.7296, 	-73.9869]])



array([3311.1669076])

We predict the price = 3311 dollars, for the first row data,  bedroom= 1, latitude=40.7296, longitude= -73.9869. It is similar the real price in the first row, 3000 dollars.  

R-squared values range from 0 to 1 and are commonly stated as percentages from 0% to 100%.  It shows how closely the data fit the regression line, so the R-squared in a simple regression is related to the correlation between the two variables. 

In [41]:
# calculate R^2 of price with bedroom, latitude, and longitude, three features.
X, y = df[['bedrooms','latitude', 'longitude']], df.price
model.fit(X, y)
model.score(X, y)


0.3690159022289894

#### So we get R squared is 0.369.

Try to plot 3d graph. So choose two features( bedrooms and latitude).

In [0]:
x= df['bedrooms']
y= df['latitude']
z = df['price']



In [0]:
import itertools
import numpy as np
import plotly.express as px
import plotly.graph_objs as go
from sklearn.linear_model import LinearRegression

def regression_3d(df, x, y, z, num=100, **kwargs):
    """
    Visualize linear regression in 3D: 2 features + 1 target
    
    df : Pandas DataFrame
    x : string, feature 1 column in df
    y : string, feature 2 column in df
    z : string, target column in df
    num : integer, number of quantiles for each feature
    """
    
    # Plot data
    fig = px.scatter_3d(df, x, y, z, **kwargs)
    
    # Fit Linear Regression
    features = [x, y]
    target = z
    model = LinearRegression()
    model.fit(df[features], df[target])    
    
    # Define grid of coordinates in the feature space
    xmin, xmax = df[x].min(), df[x].max()
    ymin, ymax = df[y].min(), df[y].max()
    xcoords = np.linspace(xmin, xmax, num)
    ycoords = np.linspace(ymin, ymax, num)
    coords = list(itertools.product(xcoords, ycoords))
    
    # Make predictions for the grid
    predictions = model.predict(coords)
    Z = predictions.reshape(num, num).T
    
    # Plot predictions as a 3D surface (plane)
    fig.add_trace(go.Surface(x=xcoords, y=ycoords, z=Z))
    
    return fig

In [48]:
regression_3d(
    train,
    x='bedrooms', 
    y='latitude', 
    z='price', 
      
)